insights/api/C/libinsights.go

```diff
@@ -47,32 +21,44 @@ import insights "github.com/ubuntu/ubuntu-insights/insights/api"
 // Otherwise, this returns NULL.
 // The error string must be freed. */
 //export collectInsights
-func collectInsights(config *C.CInsightsConfig, metricsPath *C.Cchar, flags *C.CCollectFlags) *C.char {
-	return collectCustomInsights(config, metricsPath, flags, func(conf insights.Config, metrics string, f insights.CollectFlags) error {
-		return conf.Collect(metrics, f)
+func collectInsights(config *C.CInsightsConfig, flags *C.CCollectFlags) *C.char {
+	return collectCustomInsights(config, flags, func(conf insights.Config, f insights.CollectFlags) error {
+		return conf.Collect(f)
 	})
 }
 
 // collector is a function that collects using the given parameters.
-type collector = func(conf insights.Config, metricsPath string, flags insights.CollectFlags) error
+type collector = func(conf insights.Config, flags insights.CollectFlags) error
 
 // collectCustomInsights handles C to Go translation and calls the custom collector.
-func collectCustomInsights(config *C.CInsightsConfig, metricsPath *C.Cchar, flags *C.CCollectFlags, customCollector collector) *C.char {
+func collectCustomInsights(config *C.CInsightsConfig, flags *C.CCollectFlags, customCollector collector) *C.char {
 	conf := toGoInsightsConfig(config)
 
-	mpath := ""
-	if metricsPath != nil {
-		mpath = C.GoString(metricsPath)
-	}
-
 	f := insights.CollectFlags{}
 	if flags != nil {
 		f.Period = (uint)(flags.period)
 		f.Force = (bool)(flags.force)
 		f.DryRun = (bool)(flags.dryRun)
+
+		if sourceMetricsJSONLen := (uint)(flags.sourceMetricsJSONLen); sourceMetricsJSONLen > 0 {
+			const maxJSONSize = 10 * 1024 * 1024 // 10MB limit imposed by the C API
+			if sourceMetricsJSONLen > maxJSONSize {
+				return C.CString("sourceMetricsJSONLen exceeds the maximum allowed size of 10MB")
+			}
+
+			// &flags.anon0[0] - Gets the address of the first element of the union array
+			// We then attempt to interpret the unsafe pointer of that address as specific types.
+			sourceMetricsJSON := *(*unsafe.Pointer)(unsafe.Pointer(&flags.anon0[0]))
```

Comment:
I feel like the old way was better, being able to use `sourceMetricsJSON` by name is valuable so unexpected things don't happen if the members change order or unrelated ones are added to the struct.